Adding !found condition to the for loop through sites to stop processing once resource found.

mike-guy opened the pull request:
Fixes issue #119 

Where multiple templates are associated to a site there will be multiple entries in the sites list. The for loop does not break once the resource is found (the break on line 163 has no effect at the top level - though I've left that in for now).

I've added a !found condition to the sites for loop, so that once a resource is found, the loop will stop processing and it can return to Terraform.

mike-guy commented:
The behaviour did not break the other loops, but they did iterate unnecessarily even when a resource was found/updated/deleted. To keep things consistent I've added similar logic to the other for loops.

lhercot commented:
I understand this is reducing the number of non required iterations but I am not sure it actually fixes the bug you mentioned but more avoid it. I suspect that there is a wrong condition if the additional executions overwrite something.

Could you explain a little bit more which line is overwriting in which case?

mike-guy commented:
Hi @lhercot - some more info below for you.

### MSO Config snippet
A snippet of the relevant MSO configuration below...

```json
"sites": [
    {
        "siteId": "622ef96d47b6a2762c928e20",
        "templateName": "test-template-1",
        "anps": [],
        "vrfs": [],
        "bds": [
            {
                "bdRef": "/schemas/622ef9972b0000cf003e0584/templates/test-template-1/bds/test-bd",
                "subnets": [
                    {
                        "ip": "10.1.1.1/24",
                        "description": "",
                        "scope": "private",
                        "shared": false,
                        "querier": false,
                        "virtual": false,
                        "primary": false
                    }
                ],
                "l3Outs": [],
                "hostBasedRouting": false,
                "l3OutRefs": [],
                "mac": "00:22:BD:F8:19:FF"
            }
        ],
        "contracts": [],
        "externalEpgs": [],
        "serviceGraphs": [],
        "intersiteL3outs": [],
        "networks": []
    },
    {
        "siteId": "622ef96d47b6a2762c928e20",
        "templateName": "test-template-2",
        "anps": [],
        "vrfs": [],
        "bds": [],
        "contracts": [],
        "externalEpgs": [],
        "serviceGraphs": [],
        "intersiteL3outs": [],
        "networks": []
    }
]
```

### Debug output
Snippet of the debug output with some added debug outputs added to show what is going on (explanation underneath). I've added **IMPORT ISSUE** to make them easy to see.

```
2022-03-14T08:56:40.165Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] 622ef9972b0000cf003e0584/site/622ef96d47b6a2762c928e20/bd/test-bd/ip/10.1.1.1/24: Beginning Import
2022-03-14T08:56:40.165Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 HTTP request GET mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.166Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Begin Injection
2022-03-14T08:56:40.166Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 HTTP request POST /login
2022-03-14T08:56:40.166Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 HTTP request after injection POST /login
2022-03-14T08:56:40.166Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Begining DO method https://198.18.133.100/login
2022-03-14T08:56:40.317Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 
2022-03-14T08:56:40.318Z [DEBUG] provider.terraform-provider-mso: HTTP Request: POST https://198.18.133.100/login
2022-03-14T08:56:40.318Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 nHTTP Response: 200 200 OK &{200 OK 200 HTTP/1.1 1 1 map[Content-Length:[1760] Content-Type:[application/json] Date:[Mon, 14 Mar 2022 08:56:40 GMT] Set-Cookie:[AuthCookie=eyJhbGciOiJSUzI1NiIsImtpZCI6ImExcG51dThzN3lkdTh4OW5yZnN4ZmZ6NnBjcW04bHJrIiwidHlwIjoiSldUIn0.eyJleHAiOjE2NDcyNDk0MDAsImlhdCI6MTY0NzI0ODIwMCwiaWQiOiI0OGQxMDViZGZiYzQ5YTVmY2YzOWExMGI5NjEzODZlNjFkZmU0MDA4NWNiMDMxNWQ4MThiNzYxYzU3MzVkYWZhIiwiaXNzIjoibmQiLCJyYmFjIjpbeyJkb21haW4iOiJhbGwiLCJyb2xlc1IiOjE2Nzc3MjE2LCJyb2xlc1ciOjF9XSwic2Vzc2lvbmlkIjoiQmV2eFFlUDlDSXFycHg3MVhaWFBhcmp0IiwidXNlcmZsYWdzIjowLCJ1c2VyaWQiOjI1MDAyLCJ1c2VybmFtZSI6ImFkbWluIiwidXNlcnR5cGUiOiJsb2NhbCJ9.b8phINC75aZ_n5Y2PoXMtMk23lB-tnNbY_fHd9cZItd0addcTWMjzcGlDEXtKmW-mHKYP7wW-uF-zhnSA9N5aaTSde9byMM9fJe_6_xhXhVbyEmrieXPy9f7a38bx6RY75rNNbUy8Hu4NmFhsQD1cyShjUYYjF7IeffOgkQ1sO1clVpZ_H1rcJVtu6GQuaE0VA7XsRk8gXY3i5gW1WWfMnMYl139Hvn0hA0JFUnZ1wAlUAct9_-_vZM9O9tbni1VvNBXjpWw3HKWkO0CeqWzzMI7V02amOIK9e3ui942ws8nw2pr6ln-00nM75EyvBZfq-7eyd1ehUQbgJdtVGrDxw; Max-Age=1200] Strict-Transport-Security:[max-age=31536000; includeSubDomains] X-Content-Type-Options:[nosniff] X-Xss-Protection:[1; mode=block]] 0xc000370240 1760 [] false false map[] 0xc0006ae400 0xc00042fd90}
2022-03-14T08:56:40.319Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 
2022-03-14T08:56:40.319Z [DEBUG] provider.terraform-provider-mso:  HTTP response unique string POST https://198.18.133.100/login {"jwttoken":"<removed>","username":"admin","usertype":"local","rbac":"[{\"domain\":\"all\",\"rolesR\":16777216,\"rolesW\":1}]","statusCode":200,"token":"<removed>"}
2022-03-14T08:56:40.320Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Exit from do method
2022-03-14T08:56:40.320Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 HTTP request after injection GET mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.321Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Begining DO method https://198.18.133.100/mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.386Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 
2022-03-14T08:56:40.387Z [DEBUG] provider.terraform-provider-mso: HTTP Request: GET https://198.18.133.100/mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.387Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 nHTTP Response: 200 200 OK &{200 OK 200 HTTP/1.1 1 1 map[Cache-Control:[no-cache] Content-Language:[en] Content-Length:[2139] Content-Type:[application/json] Date:[Mon, 14 Mar 2022 08:56:40 GMT] Expires:[Thu, 01 Jan 1970 00:00:01 GMT] Server:[nginx] Strict-Transport-Security:[max-age=31536000; includeSubDomains] X-Content-Type-Options:[nosniff] X-Frame-Options:[SAMEORIGIN] X-Ratelimit-Limit:[-1] X-Ratelimit-Remaining:[-1] X-Ratelimit-Reset:[1559582945] X-Xss-Protection:[1; mode=block]] 0xc0002381c0 2139 [] false false map[] 0xc0006ae300 0xc00042fd90}
2022-03-14T08:56:40.387Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 
2022-03-14T08:56:40.388Z [DEBUG] provider.terraform-provider-mso:  HTTP response unique string GET https://198.18.133.100/mso/api/v1/schemas/622ef9972b0000cf003e0584 {"id":"622ef9972b0000cf003e0584","displayName":"test-schema","description":"","templates":[{"name":"test-template-1","displayName":"test-template-1","tenantId":"622ef97c2b0000cc003e0583","anps":[],"vrfs":[{"name":"test-vrf","displayName":"test-vrf","vrfRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/vrfs/test-vrf","l3MCast":false,"preferredGroup":false,"vzAnyEnabled":false,"vzAnyProviderContracts":[],"vzAnyConsumerContracts":[],"rpConfigs":[],"pcEnfPref":"enforced","ipDataPlaneLearning":"enabled","vrfDescr":""}],"bds":[{"name":"test-bd","displayName":"test-bd","bdRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/bds/test-bd","l2UnknownUnicast":"proxy","intersiteBumTrafficAllow":false,"optimizeWanBandwidth":false,"l2Stretch":false,"subnets":[],"vrfRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/vrfs/test-vrf","unkMcastAct":"flood","v6unkMcastAct":"flood","arpFlood":true,"multiDstPktAct":"bd-flood","dhcpLabels":[],"unicastRouting":true}],"contracts":[],"filters":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"templateType":"stretched-template","templateSubType":[],"networks":[]},{"name":"test-template-2","displayName":"test-template-2","tenantId":"622ef97c2b0000cc003e0583","anps":[],"vrfs":[],"bds":[],"contracts":[],"filters":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"templateType":"stretched-template","templateSubType":[],"networks":[]}],"_updateVersion":7,"sites":[{"siteId":"622ef96d47b6a2762c928e20","templateName":"test-template-1","anps":[],"vrfs":[],"bds":[{"bdRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/bds/test-bd","subnets":[{"ip":"10.1.1.1/24","description":"","scope":"private","shared":false,"querier":false,"virtual":false,"primary":false}],"l3Outs":[],"hostBasedRouting":false,"l3OutRefs":[],"mac":"00:22:BD:F8:19:FF"}],"contracts":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"networks":[]},{"siteId":"622ef96d47b6a2762c928e20","templateName":"test-template-2","anps":[],"vrfs":[],"bds":[],"contracts":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"networks":[]}]}
2022-03-14T08:56:40.388Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Exit from do method
2022-03-14T08:56:40.388Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d schema has just been set to: 622ef9972b0000cf003e0584
2022-03-14T08:56:40.388Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** Number of sites is: 2
2022-03-14T08:56:40.388Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** Site iteration is: 0
2022-03-14T08:56:40.389Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** Site ID is: 622ef96d47b6a2762c928e20
2022-03-14T08:56:40.389Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** Inside 'if apiSite ==stateSite'
2022-03-14T08:56:40.389Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d site_id has just been set to 622ef96d47b6a2762c928e20
2022-03-14T08:56:40.389Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d template_name has just been set to test-template-1
2022-03-14T08:56:40.389Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** BD count is is: 1
2022-03-14T08:56:40.389Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** BD iteration is: 0
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** BD ref is: /schemas/622ef9972b0000cf003e0584/templates/test-template-1/bds/test-bd
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d bd_name has just been set to test-bd
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** subnet count is: 1
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** subnet iteration is is: 0
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** subnetCont var is: {"description":"","ip":"10.1.1.1/24","primary":false,"querier":false,"scope":"private","shared":false,"virtual":false}
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** State IP is: 10.1.1.1/24
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** apiIP var is: 10.1.1.1/24
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d ip has just been set to 10.1.1.1/24
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d description has just been set to 
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d scope has just been set to private
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d shared has just been set to false
2022-03-14T08:56:40.390Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d querier has just been set to false




2022-03-14T08:56:40.391Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** Site iteration is: 1
2022-03-14T08:56:40.391Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** Site ID is: 622ef96d47b6a2762c928e20
2022-03-14T08:56:40.391Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** Inside 'if apiSite ==stateSite'
2022-03-14T08:56:40.391Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d site_id has just been set to 622ef96d47b6a2762c928e20
2022-03-14T08:56:40.391Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** d template_name has just been set to test-template-2
2022-03-14T08:56:40.391Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] **IMPORT ISSUE** BD count is is: 0
2022-03-14T08:56:40.391Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] 10.1.1.1/24: Import finished successfully
2022-03-14T08:56:40.392Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] 10.1.1.1/24: Beginning Read
2022-03-14T08:56:40.392Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 HTTP request GET mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.392Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Begin Injection
2022-03-14T08:56:40.392Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 HTTP request after injection GET mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.393Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Begining DO method https://198.18.133.100/mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.456Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 
2022-03-14T08:56:40.456Z [DEBUG] provider.terraform-provider-mso: HTTP Request: GET https://198.18.133.100/mso/api/v1/schemas/622ef9972b0000cf003e0584
2022-03-14T08:56:40.457Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 nHTTP Response: 200 200 OK &{200 OK 200 HTTP/1.1 1 1 map[Cache-Control:[no-cache] Content-Language:[en] Content-Length:[2139] Content-Type:[application/json] Date:[Mon, 14 Mar 2022 08:56:40 GMT] Expires:[Thu, 01 Jan 1970 00:00:01 GMT] Server:[nginx] Strict-Transport-Security:[max-age=31536000; includeSubDomains] X-Content-Type-Options:[nosniff] X-Frame-Options:[SAMEORIGIN] X-Ratelimit-Limit:[-1] X-Ratelimit-Remaining:[-1] X-Ratelimit-Reset:[1559582945] X-Xss-Protection:[1; mode=block]] 0xc000138100 2139 [] false false map[] 0xc000378500 0xc00042fd90}
2022-03-14T08:56:40.457Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 
2022-03-14T08:56:40.457Z [DEBUG] provider.terraform-provider-mso:  HTTP response unique string GET https://198.18.133.100/mso/api/v1/schemas/622ef9972b0000cf003e0584 {"id":"622ef9972b0000cf003e0584","displayName":"test-schema","description":"","templates":[{"name":"test-template-1","displayName":"test-template-1","tenantId":"622ef97c2b0000cc003e0583","anps":[],"vrfs":[{"name":"test-vrf","displayName":"test-vrf","vrfRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/vrfs/test-vrf","l3MCast":false,"preferredGroup":false,"vzAnyEnabled":false,"vzAnyProviderContracts":[],"vzAnyConsumerContracts":[],"rpConfigs":[],"pcEnfPref":"enforced","ipDataPlaneLearning":"enabled","vrfDescr":""}],"bds":[{"name":"test-bd","displayName":"test-bd","bdRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/bds/test-bd","l2UnknownUnicast":"proxy","intersiteBumTrafficAllow":false,"optimizeWanBandwidth":false,"l2Stretch":false,"subnets":[],"vrfRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/vrfs/test-vrf","unkMcastAct":"flood","v6unkMcastAct":"flood","arpFlood":true,"multiDstPktAct":"bd-flood","dhcpLabels":[],"unicastRouting":true}],"contracts":[],"filters":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"templateType":"stretched-template","templateSubType":[],"networks":[]},{"name":"test-template-2","displayName":"test-template-2","tenantId":"622ef97c2b0000cc003e0583","anps":[],"vrfs":[],"bds":[],"contracts":[],"filters":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"templateType":"stretched-template","templateSubType":[],"networks":[]}],"_updateVersion":7,"sites":[{"siteId":"622ef96d47b6a2762c928e20","templateName":"test-template-1","anps":[],"vrfs":[],"bds":[{"bdRef":"/schemas/622ef9972b0000cf003e0584/templates/test-template-1/bds/test-bd","subnets":[{"ip":"10.1.1.1/24","description":"","scope":"private","shared":false,"querier":false,"virtual":false,"primary":false}],"l3Outs":[],"hostBasedRouting":false,"l3OutRefs":[],"mac":"00:22:BD:F8:19:FF"}],"contracts":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"networks":[]},{"siteId":"622ef96d47b6a2762c928e20","templateName":"test-template-2","anps":[],"vrfs":[],"bds":[],"contracts":[],"externalEpgs":[],"serviceGraphs":[],"intersiteL3outs":[],"networks":[]}]}
2022-03-14T08:56:40.457Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] Exit from do method
2022-03-14T08:56:40.457Z [DEBUG] provider.terraform-provider-mso: 2022/03/14 08:56:40 [DEBUG] : Read finished successfully
2022-03-14T08:56:40.458Z [WARN]  Provider "mike/local/mso" produced an unexpected new value for mso_schema_site_bd_subnet.subnet during refresh.
      - Root resource was present, but now absent
2022-03-14T08:56:40.458Z [ERROR] vertex "import mso_schema_site_bd_subnet.subnet result" error: Cannot import non-existent remote object
2022-03-14T08:56:40.459Z [ERROR] vertex "mso_schema_site_bd_subnet.subnet (import id \"622ef9972b0000cf003e0584/site/622ef96d47b6a2762c928e20/bd/test-bd/ip/10.1.1.1/24\")" error: Cannot import non-existent remote object
2022-03-14T08:56:40.462Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-03-14T08:56:40.463Z [DEBUG] provider: plugin process exited: path=.terraform/providers/mike/local/mso/0.5.0/linux_amd64/terraform-provider-mso pid=2401
2022-03-14T08:56:40.463Z [DEBUG] provider: plugin exited
```

### Explanation

1. Site iteration 0 - This is all correct and sets the value as expected. At this point, the resource is found and processing should stop. Values at this point:
    * d.schema = 622ef9972b0000cf003e0584
    * d.site_id = 622ef96d47b6a2762c928e20
    * d.template_name = template_name
    * d.bd_name = bd_name
    * d.ip = 10.1.1.1/24
    * d.description = ""
    * d.scope = "private"
    * d.shared = false
    * d.querier = false
3. Site iteration 1 happens (as there are two templates associated to the site). This results in previously correct values getting overridden, specifically:
    * Template name - now changed from "test-template-1" to "test-template-2"

### Options to fix
1. Ensure the loop at the top level has a condition for found (as per this PR). This fixes the issue and prevents unnecessary processing and stops when it should. Not to say some additional failsafes (like below) couldn't be added too.
2. Add additional logic such as checking that the template name matches the input ID. This would not fix the unnecessary iterations however, hence I preferred option one.
3. Add additional break clauses for "if found" at the various levels of the nested loops.

Hope that helps!

shrsr commented on mso/resource_mso_schema_site_bd_subnet.go:
```diff
@@ -105,7 +105,7 @@ func resourceMSOSchemaSiteBdSubnetImport(d *schema.ResourceData, m interface{})
 	stateBd := get_attribute[4]
 	stateIp := import_split[2]
-	for i := 0; i < count; i++ {
+	for i := 0; i < count && !found; i++ {
```
@mike-guy Can we forego the **!found** in the loop condition by adding the following?

`terraform import mso_schema_site_bd_subnet.sub1 {schema_id}/site/{site_id}-{template_name}/bd/{bd_name}/ip/{ip}`
```
templateInfo := regexp.MustCompile("-").Split(get_attribute[2], 2)
site_template := strings.Split(get_attribute[2], "-")
stateSite := site_template[0]
stateTemplate := templateInfo[1]
apiTemplate := models.StripQuotes(tempCont.S("templateName").String())
if apiSite == stateSite && apiTemplate == stateTemplate {
```

shrsr commented:
@lhercot Can we close this issue now?

lhercot commented:
@mike-guy Can you test the latest master to see if it fixes your issue?

lhercot commented:
We think this issue is fixed in the latest version of the provider. Please re-open if this is still an issue.
